Ensure default client mechanism is threadsafe

fjetter opened the pull request:
I encountered a problem that the default client and respective config settings were not properly reset when multiple clients were used and they were not closed in the correct order. 

For example [`test_cancel_multi_client`](https://github.com/dask/distributed/blob/b3f50ceff4012a64f8517da25060e24180c8a6d4/distributed/tests/test_client.py#L2365-L2390) closes the clients in the same order they were started but that would leave a global config set, e.g. scheduler="dask.distributed"

This led to leakage of state in test such that some tests would behave differently depending on what test was executed 

I noticed this happening in https://github.com/dask/distributed/pull/5791 but have no idea why we're not seeing this on main. I added a pytest autouse fixture to ensure no tests are leaking global clients.

Closes https://github.com/dask/distributed/issues/5772

Note: There are more issues about thread safe get_clients that are not addressed by this:  https://github.com/dask/distributed/issues/3827, https://github.com/dask/distributed/pull/5467

### Shortcoming / out of scope
If a user defines multiple default clients, modifies any of the keys `scheduler` or `shuffle` themselves, closing the last global/default client will overwrite the manual user setting to what it was before any Client was initialized. I'm fine with this since this _should_ not happen

fjetter commented:
Possibly related failure in `distributed/tests/test_worker_client.py::test_submit_different_names`

fjetter commented:
Turns out `distributed/tests/test_worker_client.py::test_submit_different_names` is easily reproducible but the error condition is still confusing and I don't understand, yet, why the issue is not 100% reproducible.

The problem is that we're still hitting https://github.com/dask/distributed/issues/2058 because https://github.com/dask/distributed/pull/2066 was only a partial fix. We're not resolving addresses everywhere and some internal checks are comparing `localhost` to `127.0.0.1` and initialize a second client 💥

graingert commented on distributed/client.py:
```diff
@@ -937,6 +942,11 @@ def __init__(
         ReplayTaskClient(self)
 
+    def _set_default_configs(self):
+        self._set_config_stack.push(
```
the `dask.config.set.__enter__` is a noop so `push` and `enter_context` do the same thing - but I think it's clearer to use `enter_context` anyway

```suggestion
        self._set_config_stack.enter_context(
```

graingert commented on distributed/client.py:
```diff
@@ -753,6 +761,7 @@ class Client(SyncMethodMixin):
 
     _default_event_handlers = {"print": _handle_print, "warn": _handle_warn}
 
+    _set_config_stack = contextlib.ExitStack()
```
this is a ClassVar and so there's a global ExitStack at `distributed.client.Client._set_config_stack` and ExitStack isn't threadsafe

graingert replied:
maybe something like this?

```python
class _ConfigureDask:
    def __init__(self):
        self._rlock = threading.RLock()
        self._count = 0

    def __enter__(self):
        with self._rlock:
            self._count += 1
            if self._count == 1:
                self._set = dask.config.set(scheduler="dask.distributed", shuffle="tasks")

    def __exit__(self, *exc_info):
        with self._rlock:
            self._count += 1
            if self._count == 0:
                self._set.__exit__(None, None, None)
                del self._set

_configure_dask = _ConfigureDask()
```

graingert replied:
or maybe something like this:


```python
class _GlobalClientManager:
    def __init__(self):
        self._lock = threading.RLock()
        self._global_clients: weakref.WeakValueDictionary[
            int, Client
        ] = weakref.WeakValueDictionary()
        self._global_client_index = 0
        self._set = None

    def _get_global_client(self) -> Client | None:
        with self._lock:
            L = sorted(self._global_clients, reverse=True)
            for k in L:
                c = _global_clients[k]
                if c.status != "closed":
                    return c
                else:
                    del _global_clients[k]
            return None

    def _set_global_client(self, c: Client) -> None:
        with self._lock:
            if not self._set:
                self._set = dask.config.set(
                    scheduler="dask.distributed", shuffle="tasks"
                )
            self._global_clients[self._global_client_index] = c
            self._global_client_index += 1

    def _del_global_client(self, c: Client) -> None:
        with self._lock:
            for k in list(self._global_clients):
                try:
                    if _global_clients[k] is c:
                        del _global_clients[k]
                except KeyError:  # pragma: no cover
                    pass

            if not self._global_clients:
                self._set.__exit__(None, None, None)
                self._set = None


_global_client_manager = _GlobalClientManager()
_set_global_client = _global_client_manager._set_global_client
_get_global_client = _global_client_manager._get_global_client
_del_global_client = _global_client_manager._del_global_client
```

and it would fix https://github.com/dask/distributed/issues/5772 also

sjperkins replied:
I'm not sure what our appetite for metaclasses in the code base is, but I've personally liked using [Multiton's](https://en.wikipedia.org/wiki/Multiton_pattern) for heavyweight resource objects which should be unique per-process for the arguments used to create it.  

```python
import weakref
from threading import Lock

class ClientMultiton(type):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.__cache = weakref.WeakValueDictionary()
        self.__lock = Lock()

    def __call__(cls, *args, **kw):
        key = args + (frozenset(kw.items()),) if kw else args

        with cls.__lock:
            try:
                # Obtain existing instance for this key
                instance = cls.__cache[key]
            except KeyError:
                pass
            else:
                # Return instance if it's not closed
                if not instance.closed:
                    return instance

                # Delete the closed cache entry
                # and proceed with recreating the instance
                # for this key
                del cls.__cache[key]

            cls.__cache[key] = instance = type.__call__(cls, *args, **kw)
            return instance


class Client(metaclass=ClientMultiton):
    def __init__(self, address=None, timeout=100):
        self.address = address
        self.timeout = timeout
        self._closed = False
        self._lock = Lock()

    @property
    def closed(self):
        # Dangerous if self._closed can transition from False to True
        with self._lock:  
            return self._closed

    def close(self):
        with self._lock:
            if self._closed:
                return

            print(f"Closing {self.address}")
            self._closed = True

    def __del__(self):
        self.close()

if __name__ == "__main__":
    def inner():
        c1 = Client(address="tcp://172.0.0.1")
        assert Client(address="tcp://172.0.0.1") is c1
        assert Client(address="tcp://172.0.0.1") is Client(address="tcp://172.0.0.1")

        addresses = list(sorted(c.address for c in Client._ClientMultiton__cache.values()))
        assert addresses == ["tcp://172.0.0.1"]

        c2 = Client(address="tcp://172.0.0.2")
        assert c2 is not c1
        addresses = list(sorted(c.address for c in Client._ClientMultiton__cache.values()))
        assert addresses == ["tcp://172.0.0.1", "tcp://172.0.0.2"]

        c1.close()
        addresses = list(sorted(c.address for c in Client._ClientMultiton__cache.values()))
        assert addresses == ["tcp://172.0.0.1", "tcp://172.0.0.2"]

        c3 = Client(address="tcp://172.0.0.1")
        assert c3 is not c1
        assert not c3.closed
        addresses = list(sorted(c.address for c in Client._ClientMultiton__cache.values()))
        assert addresses == ["tcp://172.0.0.1", "tcp://172.0.0.2"]
    
    inner()
    print("Done")
    assert len(Client._ClientMultiton__cache) == 0
```

One of the nice things about the above pattern is that one can defer to the garbage collector for cleanup by simply removing any references to the object (in a Future for instance?)

fjetter replied:
> I'm not sure what our appetite for metaclasses in the code base

Generally speaking I'm not a huge fan of metaclasses. Mostly because they are used very rarely and most people are having a hard time understanding it.

However, your suggestion looks pretty appealing and I'm intrigued. I'll look into this and would consider using the metaclass if it helps us to reduce code complexity in other areas.

sjperkins replied:
> We should keep this in mind. I think having this would help in some cases and would allow us to write a cleaner handlign of clients overall. I expect this to be a bigger effort than what I feel comfortable doing in this PR

There are also examples of a lazily initialised Multiton (with support for a finaliser) under a [BSD3 license](https://github.com/ratt-ru/dask-ms/blob/master/LICENSE) in the following locations

https://github.com/ratt-ru/dask-ms/blob/master/daskms/patterns.py
https://github.com/ratt-ru/dask-ms/blob/master/daskms/tests/test_patterns.py

graingert commented on distributed/client.py:
```diff
@@ -1121,6 +1129,8 @@ def _send_to_scheduler(self, msg):
     async def _start(self, timeout=no_default, **kwargs):
         self.status = "connecting"
+        if self._set_as_default:
+            _set_global_client(self)
```
I think moving `_set_global_client` into `async def _start` is worse - as this is likely to be called via `Client(asynchronous=False)` and so run in an off-main-thread eventloop thread

fjetter commented on distributed/tests/test_worker_client.py:
```diff
@@ -301,24 +301,41 @@ def f():
+    ],
 )
 @gen_cluster()
-async def test_submit_different_names(s, a, b):
```
Despite it's being marked flaky recently, this test has been solid before the changes proposed in this PR. This was achieved rather coincidentally since Clients were already added to the global default_client dict after initialization, before they were actually started. 
Running this test on `main` yields an exception pointing this out.

```
>   assert c_inner.status == "running"
E   AssertionError: assert 'newly-created' == 'running'
E     - running
E     + newly-created
```

What's happening on main is that

* As soon as the futures are deserialized on the worker, a
  `get_client("tcp://localhost:XXX")` is called in `Future.__setstate__`
* localhost is resolved as part of `get_client`
* since now the input address matches with the worker address, we call `Worker._get_client`
* `Worker._get_client` however does *not* match the default client (the one we created manually in the test) https://github.com/dask/distributed/blob/30f0b601ca98bbaed679ae22ce77b6ededb7ce1e/distributed/worker.py#L4296-L4306
* It will create a new client using the Worker RPC which always uses the resolved address https://github.com/dask/distributed/blob/30f0b601ca98bbaed679ae22ce77b6ededb7ce1e/distributed/worker.py#L4314-L4324
* This new client will be set to `Worker._client` *and* is the new
  `default_client`. `main` does only filter global clients for `!= closed` and
  not `running`
* Since the new one is the default client, the compute call will get the same
  one and it will always match addresses since it was created using the worker
  scheduler RPC
* This Client is initialized using `asynchronous=True` but is not awaited. Therefore, it is not actually up, yet, which is why we receive the above `newly-created` status

Diff to this branch:
- This branch requries a default client to be actually running. This leads us to
  initialize potentially many clients causing this confusion

An ideal world would...
- Detect the default client as the manually created one using localhost
- Recognize that the localhost client is talking to the same scheduler
- Not initialize any further Clients

graingert commented on distributed/client.py:
```diff
@@ -937,6 +974,11 @@ def __init__(
 
         ReplayTaskClient(self)
 
+    def _set_default_configs(self):
```
This _set_default_configs shouldn't be needed now it's managed by _GlobalClientManager

gjoseph92 commented:
I only barely skimmed this, but just wondering: how does this relate to https://github.com/dask/distributed/issues/5485 / https://github.com/dask/distributed/pull/5467#issuecomment-969611828? Basically, does this also handle asyncio correctly? Or is this is addressing an orthogonal problem? I guess I would have expected to see some contextvars in here, but I'm probably misunderstanding the problem this is solving.

fjetter commented:
>  Or is this is addressing an orthogonal problem?

It relates but I believe is a different problem. The original problem I intended to solve is not related to threading at all, actually. Tom just pointed out that there are also race conditions in a threading context which is why I introduced the global client manager with a lock.

I think the asyncio detection works reliably since we introduced the SyncMethodMixin with
https://github.com/dask/distributed/blob/2d68dfc8d29b8d0044d627aa56be9f2e94ce683e/distributed/utils.py#L316-L325

graingert commented on distributed/client.py:
```diff
@@ -108,42 +111,75 @@
+            L = sorted(self._global_clients, reverse=True)
+            for k in L:
+                c = self._global_clients[k]
+                if c.status != "closed":
```
can the client be in the "closed" status here? - `client.close` calls `_del_global_client`

graingert commented on distributed/client.py:
```diff
@@ -758,6 +796,7 @@ class Client(SyncMethodMixin):
 
     _default_event_handlers = {"print": _handle_print, "warn": _handle_warn}
 
+    _set_config_stack = contextlib.ExitStack()
```
```suggestion
```

graingert reviewed (changes requested):
still some left-over _set_config_stack

graingert commented on distributed/client.py:
```diff
@@ -1479,6 +1513,7 @@ async def _close(self, fast=False):
             return
 
         self.status = "closing"
+        _del_global_client(self)
```
might be worth avoiding the lock acquisition if we didn't _set_as_default

```suggestion
        if self._set_as_default:
            _del_global_client(self)
```

fjetter commented:
Indeed, test_package_install_restarts_on_nanny is related. It is starting a worker client and this client blocks something during shutdown
